sound/soc/sof/intel/hda-dai.c

```diff
@@ -169,19 +169,28 @@ static int hda_link_pcm_prepare(struct snd_pcm_substream *substream,
 				struct snd_soc_dai *dai)
 {
 	struct snd_soc_pcm_runtime *rtd = substream->private_data;
+	struct snd_soc_dpcm *dpcm;
 	struct snd_sof_dev *sdev =
 		snd_soc_component_get_drvdata(dai->component);
 	struct snd_sof_pcm *spcm;
 	int stream = substream->stream;
 
-	spcm = snd_sof_find_spcm_dai(sdev, rtd);
-	if (!spcm)
-		return -EINVAL;
+	for_each_dpcm_fe(rtd, stream, dpcm) {
+		spcm = snd_sof_find_spcm_dai(sdev, dpcm->fe);
+		if (!spcm)
+			return -EINVAL;
 
```

Comment:
I'm not sure EINVAL is a good errno or not. Maybe ENODEV is better? As you know, Linus is very strict with the errno.